luigi/batch_notifier.py

```diff
@@ -0,0 +1,201 @@
+
+
+def _plural_format(template, number, plural='s'):
+    if number == 0:
+        return ''
+    return template.format(number, '' if number == 1 else plural)
+
+
+class BatchNotifier(object):
+    def __init__(self, **kwargs):
+        self._config = batch_email(**kwargs)
+        self._fail_counts = collections.defaultdict(collections.Counter)
+        self._disabled_counts = collections.defaultdict(collections.Counter)
+        self._scheduling_fail_counts = collections.defaultdict(collections.Counter)
+        self._fail_expls = collections.defaultdict(_fail_queue(self._config.error_messages))
+        self._update_next_send()
+
+        self._email_format = email().format
+        if email().receiver:
+            self._default_owner = set(filter(None, email().receiver.split(',')))
+        else:
+            self._default_owner = set()
+
+    def _update_next_send(self):
+        self._next_send = time.time() + 60 * self._config.email_interval
+
+    def _key(self, task_name, family, unbatched_args):
+        if self._config.batch_mode == 'all':
+            return task_name
+        elif self._config.batch_mode == 'family':
+            return family
+        elif self._config.batch_mode == 'unbatched_params':
+            param_str = six.u(', ').join(six.u('{}={}').format(*kv) for kv in six.iteritems(unbatched_args))
+            return six.u('{}({})').format(family, param_str)
+        else:
+            raise ValueError('Unknown batch mode for batch notifier: {}'.format(
+                self._config.batch_mode))
+
+    def _format_expl(self, expl):
+        lines = expl.rstrip().split('\n')[-self._config.error_lines:]
+        if self._email_format == 'html':
+            return six.u('<pre>{}</pre>').format('\n'.join(lines))
+        else:
+            return six.u('\n{}').format(six.u('\n').join(map(six.u('      {}').format, lines)))
+
+    def _expl_body(self, expls):
+        lines = [self._format_expl(expl) for expl in expls]
+        if lines and self._email_format != 'html':
+            lines.append('')
+        return '\n'.join(lines)
+
+    def _format_task(self, task_tuple):
+        task, failure_count, disable_count, scheduling_count = task_tuple
+        counts = [
+            _plural_format('{} failure{}', failure_count),
+            _plural_format('{} disable{}', disable_count),
+            _plural_format('{} scheduling failure{}', scheduling_count),
+        ]
+        count_str = six.u(', ').join(filter(None, counts))
+        return six.u('{} ({})').format(task, count_str)
+
+    def _format_tasks(self, tasks):
+        lines = map(self._format_task, sorted(tasks, key=self._expl_key))
+        if self._email_format == 'html':
+            return six.u('<li>{}').format(six.u('\n<br>').join(lines))
+        else:
+            return six.u('- {}').format(six.u('\n  ').join(lines))
+
+    def _owners(self, owners):
+        return self._default_owner | set(owners or ())
```

Comment:
Done. It's actually already always a list, so I just have to adjust the code a little to recognize that.